invenio_db/utils.py

```diff
@@ -11,14 +11,12 @@
 
 from flask import current_app
 from sqlalchemy import inspect
-from werkzeug.local import LocalProxy
 
+from .proxies import current_db
 from .shared import db
 
-_db = LocalProxy(lambda: current_app.extensions["sqlalchemy"].db)
 
-
-def rebuild_encrypted_properties(old_key, model, properties):
+def rebuild_encrypted_properties(old_key, model, properties, db=db):
```

Comment:
What error was it giving? From your explanations, using the `.shared.db` (the now default value) will always fail, otherwise there was no need to change `invenio-oauthclient/server`. If that's the case, I would remove the default value, and release a major version because of the breaking change (in any case we should probably release major given the new required min version of Python and Flask).
See [changes](https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/changes/).

Was the error maybe related to any of the following documented changes?

> - The session is scoped to the current app context instead of the thread. This requires that an app context is active. This ensures that the session is cleaned up after every request.
> - An active Flask application context is always required to access session and engine, regardless of if an application was passed to the constructor. [#508](https://github.com/pallets-eco/flask-sqlalchemy/issues/508), [#944](https://github.com/pallets-eco/flask-sqlalchemy/issues/944)

Reply:
After testing, the error is `KeyError: <weakref at 0x10cb54c70; to 'Flask' at 0x10cb5ac70>`.
It looks like that the app is not init in the tests, because of the [custom class SQLAlchemy](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/shared.py#L31).
If you change the [fixture](https://github.com/inveniosoftware/invenio-db/blob/master/tests/conftest.py#L26) to:
```diff
db = invenio_db.db = shared.db = shared.SQLAlchemy(
+      app,
        metadata=shared.MetaData(naming_convention=shared.NAMING_CONVENTION)
    )
```
then is solves **partially** the error.
I think that we need to dig more in the error.

I have also saw this in the changes:
`SQLALCHEMY_TRACK_MODIFICATIONS is disabled by default. [#727](https://github.com/pallets-eco/flask-sqlalchemy/pull/727)`
[It looks like](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/ext.py#L79) that we want it `True` instead?

Reply:
It might be that what is happening is that when we exit from the `app.app_context`, the internal weak reference is garbage collected.

Reply:
> It looks like that the app is not init in the tests, because of the [custom class SQLAlchemy](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/shared.py#L31).

I thought that the db is initialized [with](https://github.com/inveniosoftware/invenio-db/blob/9ea8eb3a6f05925d26dacf76645610bccef4f278/invenio_db/ext.py#L83) by calling it [with](https://github.com/inveniosoftware/invenio-db/blob/9ea8eb3a6f05925d26dacf76645610bccef4f278/tests/test_utils.py#L39)

> the now default value) will always fail

Do you mean because of the same name of the imported and the parameter name?
It is working on my local machine and it is also not failing for postgresql on github. I have no idea what the problems with mysql8 are!

> If you change the [fixture](https://github.com/inveniosoftware/invenio-db/blob/master/tests/conftest.py#L26) to:

how do you get the `app into the db() fixture to initialize it in the custom SQLAlchemy class?

> I think that we need to dig more in the error.

I will have a look at your pointed out documented changes. It could be that this could be the root of the problem. 

> It might be that what is happening is that when we exit from the app.app_context, the internal weak reference is garbage collected.

I tried to call `InvenioDB(app, entry_point_group=False, db=db)` within the `with app.app_context()` context. The result was the same error.

> I think that we need to dig more in the error.

I will investigate deeper into the garbage collector idea and the documented changes. Hopefully it is not necessary to change the `rebuild_encrypted_properties` function

> [It looks like](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/ext.py#L79) that we want it True instead?

the linked line sets this variable to True? I think i am missing something.

Reply:
> how do you get the `app into the db() fixture to initialize it in the custom SQLAlchemy class?
Just to test, I tried with adding the `app` fixture in the `db` fixture:
```
def db(app):
```

> the linked line sets this variable to True? I think i am missing something.
it looks to me that we are doing a `setdefault`: if it is already `False`, it won't be changed. To be tested!